src/controllers/controllermanager.cpp

```diff
@@ -168,13 +172,21 @@ void ControllerManager::updateControllerList() {
     locker.unlock();
 
     QList<Controller*> newDeviceList;
-    foreach (ControllerEnumerator* pEnumerator, enumerators) {
+    for (ControllerEnumerator* pEnumerator: enumerators) {
         newDeviceList.append(pEnumerator->queryDevices());
     }
 
+    // Since there is only one KeyboardController, it is not
+    // enumerated by an enumerator but added directly here.
+    KeyboardController* kbdController = new KeyboardController(m_pKeyboard);
+    newDeviceList.prepend(kbdController);
```

Comment:
No, I removed the KeyboardEnumerator. This is the very only place where KeyboardController is instantiated.

But wait, now that you are mentioning it. What if `newDeviceList == m_controllers`. We should probably delete `kbdController` in the else, right?

```c++
if (newDeviceList != m_controllers) {
  m_controllers = newDeviceList;
  // etc.
} else {
  delete kbdController;
}
```